pkg/git/git.go

```diff
@@ -0,0 +1,78 @@
+// Copyright The Shipwright Contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package git
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/shipwright-io/build/pkg/ctxlog"
+	"regexp"
+
+	gogitv5 "github.com/go-git/go-git/v5"
+)
+
+const (
+	defaultRemote = "origin"
+	privateRepo   = "private"
+	publicRepo    = "public"
+	invalidRepo   = "invalid"
+)
+
+var (
+	// URLRegexFormat matches the url of git repository
+	httpURLRegexFormat        = regexp.MustCompile(`^((http:\/\/)|(https:\/\/))?([a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}(\/)`)
+	sshURLRegexFormat         = regexp.MustCompile(`(ssh:\/\/[\w\d\.]+|.+@?.+\..+:)(:[\d]+){0,1}\/*(.*)`)
+	httpsPublicURLRegexFormat = regexp.MustCompile(`(https:\/\/)?(github\.)+com(\/)`)
+	sshPublicURLRegexFormat   = regexp.MustCompile(`(git+@github.com:)(:[\d]+){0,1}/*(.*)`)
+)
+
+// ValidateGitURLExists validate if a source URL exists or not
+func ValidateGitURLExists(ctx context.Context, urlPath string) error {
+	repoKind := ValidateGitURLFormat(urlPath)
+	switch repoKind {
+	case publicRepo:
+		repo := gogitv5.NewRemote(memory.NewStorage(), &config.RemoteConfig{
+			Name: defaultRemote,
+			URLs: []string{urlPath},
+		})
+		// Note: When the urlPath is an valid public path, however, this path doesn't exist, func will return `authentication required`, this maybe mislead
+		// So convert this error message to `repository not found`
+		_, err := repo.List(&gogitv5.ListOptions{})
+		if err != nil && err == transport.ErrAuthenticationRequired {
+			ctxlog.Error(ctx, err, fmt.Sprintf("repository %s not found", urlPath))
+			return fmt.Errorf("repository not found")
+		} else if err != nil {
+			ctxlog.Error(ctx, err, fmt.Sprintf("sourceURL is %s", urlPath))
+			return err
+		}
+		return nil
+	case invalidRepo:
+		ctxlog.Info(ctx, fmt.Sprintf("repository %s is invalid", urlPath))
+		return fmt.Errorf("invalid source url")
+	case privateRepo:
+		// Skip verify private repo
+		// TODO: check private repo in future
+		ctxlog.Info(ctx, fmt.Sprintf("skip check private repository %s", urlPath))
+		return nil
+	default:
+		ctxlog.Error(ctx, fmt.Errorf("unknown repository kind"), urlPath)
+		return fmt.Errorf("unknown repository kind %s", urlPath)
+	}
+}
+
+// ValidateGitURLFormat validates the given URL format is valid and public or not
+func ValidateGitURLFormat(url string) string {
```

Comment:
From what I can see, this function is not really validating anything, it is parsing the url and returning whether it is a public, private, or invalid repository, so the function name seems a bit off. How about something like "IdentifyGitURLType" or "ParseGitURLType" ?

Reply:
Done!